### PR TITLE
Limit page depth of searches

### LINKED ISSUE #857: Limit page depth of searches
## Problem

As a temporary fix for some performance issues with the Openverse API, we should cap the max page depth to `20`. This means that if a search had 40 pages of results users would still be limited to accessing pages 1-20.  

## Description

This prevents the scenario described here https://github.com/cc-archive/cccatalog-api/issues/565, which has become an increasing problem with the number of users scraping Openverse's dataset, while we make improvements to pagination and dead link elimination.

## Implementation

- [ ] 🙋 I would be interested in implementing this feature.

### PR DESCRIPTION
**Pull request by AetherUnbound**
## Fixes

Fixes #857 by @zackkrida

## Description

This PR adds a cap to the number of pages returned in API results to 20 (defined by `MAX_TOTAL_PAGE_COUNT` in `pagination.py`). It does so by:

1. Preventing the pagination class from setting a page size that's higher (e.g. when `page` is in the query parameters)
2. Limiting the number returned when determining "result and page count"

The former is pretty straightforward and I've added tests for that case. The latter makes way less sense to me. I've capped the `page_count` within that function to `MAX_TOTAL_PAGE_COUNT` as well, but the logic within it is confusing. I wrote tests along with it with some of my own notes. I hesitate to change _more_ functionality because the current calculations are "working", I only altered what the possible cap was.

## Testing Instructions

- `just api-test`
- Try http://localhost:8000/v1/images/ before and after checking this branch out and see that even though there are at least 250 pages worth of results, only 20 pages are shown (and going to page 21 and above returns an error)

## Checklist

- [x] My pull request has a descriptive title (not a vague title like `Update index.md`).
- [x] My pull request targets the *default* branch of the repository (`main`) or a parent feature branch.
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added or updated tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.

[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: api/catalog/api/utils/pagination.py [hunk 1]
```diff
@@ -42,6 +45,8 @@ def page(self, value):
         value = int(value)  # convert str params to int
         if value <= 0:
             raise get_api_exception("Page must be greater than 0.", 400)
+        elif value > 20:
+            raise get_api_exception("Searches are limited to 20 pages.", 400)
```
Comment (sarayourfriend) [inline api/catalog/api/utils/pagination.py:49]: This validation should be removed from here and the new validation should be added to the serializer: https://github.com/WordPress/openverse-api/blob/d26df92f04120e7382d042ae0b439fb6ea9209be/api/catalog/api/serializers/media_serializers.py#L167-L174

```suggestion
```

It was erroneously added back to this area during the revert.

File: api/Pipfile [hunk 2]
```diff
@@ -7,6 +7,7 @@ verify_ssl = true
 ipython = "*"
 pycodestyle = "*"
 pytest-django = ">=3.5"
+pytest-raises = "*"
```
Comment (sarayourfriend) [inline api/Pipfile:10]: Is this new dependency being used at all?

**Review by sarayourfriend — CHANGES_REQUESTED** [review]
LGTM but we should keep the validation in the serializer and remove the code added back in the revert.

File: api/test/unit/controllers/test_search_controller.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+from catalog.api.utils.pagination import MAX_TOTAL_PAGE_COUNT
+
+
+@pytest.mark.parametrize(
+    "total_hits, real_result_count, page_size, expected",
+    [
+        # No results
+        (0, 0, 10, (0, 0)),
+        # Setting page size to 0 raises an exception
+        pytest.param(
```
Comment (obulat) [inline api/test/unit/controllers/test_search_controller.py:15]: Does it make sense to add a negative page size, too?

**Comment by sarayourfriend** [comment]
Assigning this to myself as Madison is the MSR runner for this week and I will be taking this off her hands.

File: api/catalog/api/utils/pagination.py (hunk 1 above)
Comment (sarayourfriend) [reply api/catalog/api/utils/pagination.py:49]: Actually, this is inaccurate. The validation needs to be moved to the serializer, but I think it should be done in a different PR dedicated to a small refactor of the pagination parameter validation. We're under-utilising serializer validation at the moment and we can delete a significant amount code. But it should be done in a different PR.

**Review by sarayourfriend — APPROVED** [review]
LGTM.

File: api/test/unit/controllers/test_search_controller.py (hunk 3 above)
Comment (sarayourfriend) [reply api/test/unit/controllers/test_search_controller.py:15]: Upstream validation catches negative values here.

File: api/catalog/api/utils/pagination.py (hunk 1 above)
Comment (zackkrida) [inline api/catalog/api/utils/pagination.py:49]:
```suggestion
            raise get_api_exception("Searches are limited to the first 20 pages of results while we mitigate API performance issues. We are working on a solution for bulk exports of Openverse's data.", 400)
```

Might want to be super verbose here? Not sure what else we could say 🤷 CC @WordPress/openverse-maintainers if anyone else has ideas. Would we, for example, want to tell them to email us or anything? Not sure.

Comment (AetherUnbound) [reply api/catalog/api/utils/pagination.py:49]: Maybe we could just say `"Searches are currently limited to 20 pages"`, which implies that might change in the future?

File: api/Pipfile (hunk 2 above)
Comment (AetherUnbound) [reply api/Pipfile:10]: It is used here: https://github.com/WordPress/openverse-api/pull/859/files#diff-3f72d9dc992edb46dcea0b7136c9ecaa0f888653a2c4c6261ba0116de468eef8R16

**Pull request merged**